### PR TITLE
Update user search functionality

### LINKED ISSUE #142: Update search functionality for following a new user
**User statement**
*As a user, I would like to see search results if I'm unsure of the full username.

**Describe the solution you'd like**
Implement some sort of partial search autocomplete functionality according to the Figma mockup.

**Development notes**

### PR DESCRIPTION
**Pull request by JacobNguyen71**
This PR updates the user search functionality by allowing users to search for someone they wish to follow and see if that user exists by looking at the auto complete results.
Closes #142

- [x]  run unit tests
- [x]  run UI tests

When user enters no username, error message still appears
CMPUT301HabitSmasher#142 1

When the user begins searching, if the username begins with what the user searches, it will appear in the autocomplete results. User can also click on the username in the auto complete which will fill the AutoCompleteTextView
CMPUT301HabitSmasher#142 2

If User enters invalid user name, error message still aprears
CMPUT301HabitSmasher#142 3

When User tries to follow themself, error message still appears
CMPUT301HabitSmasher#142 4

Another search example
CMPUT301HabitSmasher#142 5

When the user presses follow, toast message still appears
CMPUT301HabitSmasher#142 6

### CODE REVIEW

File: app/src/main/java/com/example/habitsmasher/FollowUserDialog.java [hunk 1]
```diff
@@ -47,9 +50,10 @@
     private static final String FOLLOWERS_FIELD = "followers";
 
     private FirebaseFirestore _db;
-    private EditText _userToFollow;
+    private AutoCompleteTextView _userToFollowACTV;
     private Handler _handler;
     private Context _context;
+    private ArrayList<String> _usernames;
```
Comment (JacobNguyen71) [inline app/src/main/java/com/example/habitsmasher/FollowUserDialog.java:56]: ArrayList which will store auto complete results

File: app/src/main/java/com/example/habitsmasher/FollowUserDialog.java [hunk 2]
```diff
@@ -68,10 +72,35 @@ public View onCreateView(@NonNull LayoutInflater inflater,
...
+                .get()
+                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<QuerySnapshot> task) {
+                if (task.isSuccessful()) {
+                    for (QueryDocumentSnapshot document : task.getResult()) {
+                        if (document.exists()) {
+                            if (document.get(USERNAME_FIELD).toString().contains(_userToFollowACTV.getText().toString())) {
+                                Log.d(TAG, "Username exists");
+                                _usernames.add(document.get(USERNAME_FIELD).toString());
```
Comment (JacobNguyen71) [inline app/src/main/java/com/example/habitsmasher/FollowUserDialog.java:91]: if username in database contains what the user searches, add to the ArrayList

File: app/src/main/java/com/example/habitsmasher/FollowUserDialog.java [hunk 3]
```diff
@@ -68,10 +72,35 @@ public View onCreateView(@NonNull LayoutInflater inflater,
...
+                            }
+                        }
+                    }
+                } else {
+                    Log.d(TAG, "Error getting documents: ", task.getException());
+                }
+            }
+        });
+
+        ArrayAdapter<String> adapter = new ArrayAdapter<String>(getContext(), android.R.layout.simple_list_item_1, _usernames);
```
Comment (JacobNguyen71) [inline app/src/main/java/com/example/habitsmasher/FollowUserDialog.java:101]: Adapter for AutoCompleteTextView

**Review by jaskim9824 — DISMISSED** [review]

**Review by rudy-patel — CHANGES_REQUESTED** [review]
Nice work Jacob! Looks awesome, I'm glad to see that it wasn't a lot of work to pull this off. I just left a task regarding an edge case that came to mind when looking over the PR. 

Also, let's ensure all existing UI tests pass **and** add a UI test to test the partial search. This can be as simple as you typing in `testUs` and clicking on the `TestUser` username that pops up for our test user. No need to actually hit the follow button, just making sure that the username pops up and we can select it.

- [ ] run UI tests (if you haven't already)
- [ ] add new UI test to test partial search of a user

After that, this is good to be merged in!

File: app/src/main/java/com/example/habitsmasher/FollowUserDialog.java (hunk 2 above)
Comment (JacobNguyen71) [reply app/src/main/java/com/example/habitsmasher/FollowUserDialog.java:91]: Right now, when we query the database for users' usernames, it will not be case sensitive to what the user enters. Attached a screenshot to show it works correctly.
CMPUT301HabitSmasher#142 7

**Review by JuliePilz — DISMISSED** [review]
Great job this looks awesome!

**Review by jaskim9824 — COMMENTED** [review]
Looks good to me! Just get those merge conflicts resolved and I'll approve!

File: app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java [hunk 4]
```diff
@@ -2084,6 +2085,34 @@ public void forgotPassword_validEmailNotRegistered_emailNotSent() {
...
+        // click on the auto complete text view
+        _solo.clickOnView(_solo.getView(R.id.auto_complete_text_view));
+
+        // enter part of the users username we wish to follow
+        _solo.enterText(_solo.getEditText("Username"), TEST_USER_SUBSTRING);
+
+        // click on the user name that we wish to follow
+        _solo.clickOnText(TEST_USER_USERNAME);
+
+        assertTextOnScreen(TEST_USER_USERNAME);
```
Comment (rudy-patel) [inline app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java:2113]: Instead of asserting what text is on screen, can we assert on the value of the text view? After you click on the username from the list, the autocomplete text view should have the value of the username you just clicked.

**Review by jaskim9824 — APPROVED** [review]

File: app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java [hunk 5]
```diff
@@ -2152,6 +2152,38 @@ public void forgotPassword_validEmailNotRegistered_emailNotSent() {
...
+        // click on the auto complete text view
+        _solo.clickOnView(_solo.getView(R.id.auto_complete_text_view));
+
+        // enter part of the users username we wish to follow
+        _solo.enterText(_solo.getEditText("Username"), TEST_USER_SUBSTRING);
+
+        // click on the user name that we wish to follow
+        // Click on text does not work for item in autoCompleteTextView
+        // so I decided to go with click on screen at a specific location
+        _solo.clickOnScreen(540, 1170);
```
Comment (rudy-patel) [inline app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java:2180]: Setting hardcoded X and Y values in a UI test isn't a good approach. Sure this may pass on your emulator, but it could fail on other devices. I see the comment you've left, but this is just asking for a flaky test. I found this [link](https://stackoverflow.com/questions/19860782/how-to-test-for-autocompletetextview-itemclick-event-using-robotium). You can click on the text, you just have to wait for it to appear first.

- [ ] remove hardcoded numbers from UI test (if possible)

**Review by rudy-patel — CHANGES_REQUESTED** [review]
Just one small thing that you should change in your UI test! Other than that, it looks good!

File: app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java (hunk 5 above)
Comment (JacobNguyen71) [reply app/src/androidTest/java/com/example/habitsmasher/MainActivityTest.java:2180]: I tried this approach as well and I found it wouldn't pass the test so I ended up not using it

**Review by kaden-dreger — APPROVED** [review]
Looks good!

**Review by rudy-patel — APPROVED** [review]
